src/Data/YAML/Event/Writer.hs

```diff
@@ -124,58 +123,115 @@ putNode = \docMarker -> go (-1 :: Int) (not docMarker) BlockIn
     go _  _ _  [] _cont = error ("putNode: expected node-start event instead of end-of-stream")
     go !n !sol c (t : rest) cont = case t of
         Scalar        anc tag sty t' -> goStr (n+1) sol c anc tag sty t' (cont rest)
-        SequenceStart anc tag sty    -> goSeq (n+1) sol c anc tag sty rest cont
-        MappingStart  anc tag sty    -> goMap (n+1) sol c anc tag sty rest cont
+        SequenceStart anc tag sty    -> goSeq (n+1) sol (chn sty) anc tag sty rest cont
+        MappingStart  anc tag sty    -> goMap (n+1) sol (chn sty) anc tag sty rest cont
         Alias a                      -> pfx <> goAlias c a (cont rest)
 
         _ -> error ("putNode: expected node-start event instead of " ++ show t)
-      where -- TODO flow
+
+      where
         pfx | sol           = mempty
             | BlockKey <- c = mempty
+            | FlowKey  <- c = mempty
             | otherwise     = T.B.singleton ' '
 
+        chn sty
+          | Flow <-sty, (BlockIn == c || BlockOut == c) = FlowOut
+          | otherwise = c
 
-    goMap n sol c anc tag sty (MappingEnd : rest) cont = pfx $ "{}\n" <> cont rest
+
+    goMap _ sol _ anc tag _ (MappingEnd : rest) cont = pfx $ "{}\n" <> cont rest
       where
         pfx cont' = (if sol then mempty else ws) <> anchorTag'' (Right ws) anc tag cont'
 
-    goMap n sol c anc tag sty xs cont = case c of
-        BlockIn | not (not sol && n == 0) -- avoid "--- " case
-           ->  (if sol then mempty else ws) <> anchorTag'' (Right (eol <> mkInd n)) anc tag
-               (putKey xs (\ys -> go n False BlockOut ys g))
-        _  ->  anchorTag'' (Left ws) anc tag $ T.B.singleton '\n' <> g xs
+    goMap n sol c anc tag sty xs cont = case sty of
+      Block ->
+        case c of
+          BlockIn | not (not sol && n == 0) -- avoid "--- " case
+             ->  (if sol then mempty else ws) <> anchorTag'' (Right (eol <> mkInd n)) anc tag
```

Comment:
I notice the fragment `if sol then mempty else ws` occuring several times; maybe we should invent a small helper

```hs
wsSol :: Bool -> ...
wsSol sol = if sol then mempty else ws
```

then it'd allow to write the line

```hs
->  (if sol then mempty else ws) <> anchorTag'' (Right (eol <> mkInd n)) anc tag
```

more concisely as

```hs
->  wsSol sol <> anchorTag'' (Right (eol <> mkInd n)) anc tag
```